activesupport/lib/active_support/json/as_json_encoder.rb

```diff
@@ -0,0 +1,201 @@
+module ActiveSupport
+  module JSON
+    class AsJSONEncoder
+      def self.encode(object, options)
+        new(options).encode object
+      end
+
+      def self.encode_next(object, options)
+        new(options).visit object
+      end
+
+      def self.internal_as_json?(object)
+        object.respond_to?(:as_json) &&
+          object.method(:as_json).owner == ActiveSupport::AsJSON
+      end
+
+      def initialize(as_json_options)
+        @as_json_options = as_json_options
+      end
+
+      def encode(object)
+        if AsJSONEncoder.internal_as_json?(object)
+          visit object
+        else
+          encode object.as_json
```

Comment:
My first thought is we might as well save the direct recursion and move this up as:

```ruby
object = object.as_json until AsJSONEncoder.internal_as_json?(object)
```

... but that does mean `def as_json; self; end` goes from being a stack overflow to an iloop, so maybe it's better as is. 🤷‍♂️